### ISSUE TITLE
AWS S3 as backend

### ISSUE THREAD

**Issue by davitb** [hashicorp/vault#120]
I'm wondering why S3 is not supported as a backend option. It would provide HA and many organizations are already using S3.
Is there something I'm missing?

**Comment by mitchellh** [comment]
Simply because no one has implemented it yet. If you'd like to take a stab then that'd be great!

I don't think it can support Vault HA mode though, since it doesn't provide leader election. But the storage itself would be highly available, yes.

**Comment by davitb** [comment]
Ok, will look more into how HA is implemented in Vault. It's a bit concerning though that Consul is a prerequisite for Vault deployment.

**Comment by mitchellh** [comment]
@davitb The prerequisite is simply leader election capabilities. Consul was easiest given our background and the deployments of most of our users, but we're definitely not against any others coming in.

**Comment by davitb** [comment]
Thanks for getting back quickly. Btw, great job with Vault. Love it so far...

**Comment by carldanley** [comment]
If s3 is not supported, does backend "file" suffice for now? If so, is there an example of this configuration somewhere?

**Comment by carldanley** [comment]
Also, if "file" *did* work, couldn't me essentially move those physical files to s3 at a later point?

**Comment by nrocine** [comment]
FYI, there is a file-like backend "path". There isn't much to configure it: http://vaultproject.io/docs/config/index.html#path except that is implemented as directory, not a single file.

In messing around with vault, I've been using it as the default backend like so

```
backend "file" {
  path = "/mnt/vault_db"
}

listener "tcp" {
  address = "0.0.0.0:8200"
  tls_cert_file = "/mnt/foo.public.pem"
  tls_key_file = "/mnt/foo.private.pem"
}
```

My initial configuration has /mnt as an external EBS volume, and then the plan is to take EBS snapshots.

Once vault is inited and unsealed, /mnt/vault_db looks like so
```
[ec2-user@ip-172-31-2-68 mnt]$ find /mnt/vault_db/
/mnt/vault_db/
/mnt/vault_db/core
/mnt/vault_db/core/_seal-config
/mnt/vault_db/core/_auth
/mnt/vault_db/core/_audit
/mnt/vault_db/core/_mounts
/mnt/vault_db/barrier
/mnt/vault_db/barrier/_init
/mnt/vault_db/sys
/mnt/vault_db/sys/token
/mnt/vault_db/sys/token/id
/mnt/vault_db/sys/token/id/_6a0e95ac47e9c200e7f8e7416bfd7cda9ea44df9
/mnt/vault_db/sys/token/_salt
```

**Comment by certifiedloud** [comment]
+1, I'd love to see this feature working in Vault.

**Comment by armon** [comment]
Fixed by #242
